crates/milli/src/search/new/geo_sort.rs

```diff
@@ -267,24 +265,79 @@ impl<'ctx, Q: RankingRuleQueryTrait> RankingRule<'ctx, Q> for GeoSort<Q> {
                 cache.pop_back()
             }
         };
-        while let Some((id, point)) = next(&mut self.cached_sorted_docids) {
-            if geo_candidates.contains(id) {
-                return Ok(Some(RankingRuleOutput {
-                    query,
-                    candidates: RoaringBitmap::from_iter([id]),
-                    score: ScoreDetails::GeoSort(score_details::GeoSort {
-                        target_point: self.point,
-                        ascending: self.ascending,
-                        value: Some(point),
-                    }),
-                }));
+        let put_back = |cache: &mut VecDeque<_>, x: _| {
+            if ascending {
+                cache.push_front(x)
+            } else {
+                cache.push_back(x)
             }
-        }
+        };
 
-        // if we got out of this loop it means we've exhausted our cache.
-        // we need to refill it and run the function again.
-        self.fill_buffer(ctx, &geo_candidates)?;
-        self.next_bucket(ctx, logger, universe)
+        let mut current_bucket = RoaringBitmap::new();
+        // current_distance stores the first point and distance in current bucket
+        // The farthest distance between two points on earth is about 2e7 meters, u32 is big enough to hold any distance.
+        let mut current_distance: Option<([f64; 2], u32)> = None;
```

Comment:
In fact, distance can be represented by float. I originally wanted to use distance as the key of a map and then use u32. Now using u32 is a mistake.

Reply:
Yes why not.
But when comparing floats do not forget that it’s recommended to run `left - right < f64::EPSILON` instead of a normal `==`.